### ISSUE TITLE
Batch norm example?

### ISSUE THREAD

**Issue by xyzsam** [01org/mkl-dnn#167]
Hi,

I'm having trouble implementing batch norm for forward inference. I've tried many things and they all lead to a `mkldnn::error` exception being thrown at some point or another, or incorrect results. I think I am not getting the various dimensions or formats of the memory descriptors correct.

Based on reading the documentation, I inferred that for a FC layer, we have to compress all the dimensions for a single input into the channel dimension. That seems to be what the batch norm descriptor expects too, since if I try any other combination, I get an error with the expected sizes of the memories being different.  Let me know if this is wrong.

Here's basically what I'm trying: https://gist.github.com/xyzsam/f03124689ffbd92c03eb3ede7fab0a25

The `mkldnn::error` gets thrown at the very end when I create the `batch_normalization_forward` primitive. But I don't get any other information about what caused it. In general, it would be really useful if these errors could get propagated up. 

If I take out the `use_global_stats` flag and remove the mean and variance memories, it at least runs to completion ~~, but the output is wrong~~ (never mind that part).

**Comment by xyzsam** [comment]
It seems like when I construct the `batch_normalization_forward` object, it doesn't get the right constructor.

The one I want is:
```
    batch_normalization_forward(const primitive_desc &aprimitive_desc,
            const primitive::at &src, const primitive::at &mean,
            const primitive::at &variance, const primitive::at &weights,
            const memory &dst) {
        mkldnn_primitive_t result;
        mkldnn_primitive_at_t inputs[] = { src.data,
            mean.data, variance.data, weights.data };
        const_mkldnn_primitive_t outputs[] = { dst.get() };
        error::wrap_c_api(mkldnn_primitive_create(&result,
                aprimitive_desc.get(), inputs, outputs),
            "could not create a batch normalization forward primitive");
        reset(result);
    }  
```
The one it actually goes to is:
```
    batch_normalization_forward(const primitive_desc &aprimitive_desc,
            const primitive::at &src, const primitive::at &weights,
            const memory &dst, const memory &mean, const memory &variance) {
        mkldnn_primitive_t result;
        mkldnn_primitive_at_t inputs[] = { src.data, weights.data };
        const_mkldnn_primitive_t outputs[] = { dst.get(),
            mean.get(), variance.get() };
        error::wrap_c_api(mkldnn_primitive_create(&result,
                aprimitive_desc.get(), inputs, outputs),
            "could not create a batch normalization forward primitive");
        reset(result);
    } 
```

**Comment by emfomenk** [comment]
Hi @xyzsam,

Alas there is minor confusion with batch norm here...
We probably use overloading too much :)
You need to help compiler understand that first 5 objects are actually `primitive::at`, and not just `memory`.
The following patch fixes the issue:

``` diff
diff --git a/x.cpp b/x.cpp
index dc4e280..672d4f8 100644
--- a/x.cpp
+++ b/x.cpp
@@ -53,9 +53,9 @@ void batch_norm() {
     /* Repeat for all the other inputs */

     auto bn = batch_normalization_forward(bn_pd,
-                                          input_memory,
-                                          bn_mean_memory,
-                                          bn_var_memory,
-                                          bn_weights_memory,
+                                          primitive::at(input_memory),
+                                          primitive::at(bn_mean_memory),
+                                          primitive::at(bn_var_memory),
+                                          primitive::at(bn_weights_memory),
                                           bn_dst_memory);
 }
```

**Comment by vpirogov** [comment]
Would you mind sharing details on compiler you are using?

**Comment by vpirogov** [comment]
My question is actually irrelevant. I misunderstood @emfomenk's comment.

**Comment by xyzsam** [comment]
Thanks a lot for the help! I'm finally able to get past the exception and get the correct results. Hopefully you guys can resolve this overloading issue in the future. I'm not entirely sure why `memory` can be implicitly converted to `primitive:at`, so some documentation explaining this would be helpful. Also, it would very helpful to explicitly state what the dimensionality of the inputs are expected to be - I had to infer that from the LaTeX equation, which isn't ideal :)
